app.py

```diff
@@ -11,21 +11,19 @@ def validate(keyboard, pin):
     :param pin: a string of digits
     :return: nothing or raises ValueError if validation fails
     """
-    if not all([isinstance(keyboard, list), isinstance(pin, str)]):
+    if not all([isinstance(keyboard, list), isinstance(pin, str), keyboard != []]):
```

Comment:
Protecting against invalid input is the only way to survive in production. 👍🏾 

Suggestion: All validations worth to be separated and raise the correct exception, as follows:
- if pin is not string, `raise TypeError(msg)`
- if keyboard is not list, `raise TypeError(msg)`
- if keyboard is empty, `raise ValueError(msg)`

Reply:
Maybe we can reject an empty pin also.